// @vlayer/sdk/dist/api/prover/v_getProofReceipt.js
import {} from "../lib/types/vlayer.js";
import { handleProverResponseError, handleAuthErrors, } from "../lib/handleErrors.js";
import { InvalidProverResponseError } from "../lib/errors.js";
import { proofReceiptSchema } from "../lib/types/vlayer.js";
import { validateJrpcResponse } from "../lib/jrpc.js";
import debug from "debug";
const log = debug("vlayer:v_getProofReceipt");
function v_getProofReceiptBody(hash) {
    return {
        method: "v_getProofReceipt",
        params: { hash },
        id: 1,
        jsonrpc: "2.0",
    };
}
export async function v_getProofReceipt(hash, url = "http://127.0.0.1:3000", token) {
    const headers = {
        "Content-Type": "application/json",
    };
    if (token !== undefined) {
        headers["Authorization"] = "Bearer " + token;
    }
    const rawResponse = await fetch(url, {
        method: "POST",
        body: JSON.stringify(v_getProofReceiptBody(hash)),
        headers,
    });
    log("raw response: ", rawResponse);
    const responseJson = await rawResponse.json();
    log("response body: ", responseJson);
    if (!rawResponse.ok) {
        throw handleAuthErrors(rawResponse.status, responseJson);
    }
    const response = validateJrpcResponse(responseJson);
    if (response.error !== undefined) {
        throw handleProverResponseError(response.error);
    }
    const result = proofReceiptSchema.safeParse(response.result);
    if (!result.success) {
        throw new InvalidProverResponseError("v_getProofReceipt", response.result);
    }
    return result.data;
}
