// @vlayer/sdk/dist/api/lib/types/vlayer.js
import {} from "../../../web-proof-commons/utils.js";
import {} from "viem";
import {} from "./webProofProvider.js";
import {} from "./viem.js";
import { z } from "zod";
export const callHashSchema = z.string().startsWith("0x").length(66);
export var ProofState;
(function (ProofState) {
    ProofState["Queued"] = "queued";
    ProofState["AllocateGas"] = "allocate_gas";
    ProofState["Preflight"] = "preflight";
    ProofState["Proving"] = "proving";
    ProofState["Done"] = "done";
})(ProofState || (ProofState = {}));
export const proofReceiptSchema = z.discriminatedUnion("status", [
    z.object({
        status: z.literal(0),
        error: z.string(),
        data: z.undefined(),
        metrics: z.custom(),
        state: z.enum([
            ProofState.AllocateGas,
            ProofState.Preflight,
            ProofState.Proving,
        ]),
    }),
    z.object({
        status: z.literal(1),
        error: z.undefined(),
        state: z.enum([
            ProofState.Done,
            ProofState.AllocateGas,
            ProofState.Preflight,
            ProofState.Proving,
            ProofState.Queued,
        ]),
        data: z.custom(),
        metrics: z.custom(),
    }),
]);
export const versionsSchema = z.object({
    call_guest_id: z.string(),
    chain_guest_id: z.string(),
    api_version: z.string(),
    risc0_version: z.string(),
});
